### ISSUE TITLE
Fix S1125 FP: Type check with System.Object

### ISSUE THREAD

**Issue by kimpenhaus** [SonarSource/sonar-dotnet#7792]
### Description

I'm not sure if this is by intent but I think there is an issue with recognizing [S1125 `Boolean literals should not be redundant`](https://sonarsource.atlassian.net/browse/RSPEC-1125) when source type is not bool.

### Repro steps

```
var exception = new Exception();
exception.Data.Add("handled", true);

if (exception.Data.Contains("handled") && exception.Data["handled"] is true)
{
    // do something
}
```

`Data` is of type IDictionary - which has kay/value as object. Therefore I think `is true` pattern matching is valid at this point. If I am wrong please correct me :)

<img width="1379" alt="image" src="https://github.com/SonarSource/sonar-dotnet/assets/575109/dd8dda5f-245f-401b-bea0-d9128ac5a3bd">

Removing the pattern matching results in code that can not be compiled

<img width="1302" alt="image" src="https://github.com/SonarSource/sonar-dotnet/assets/575109/59c70e1c-cf5b-4b42-837a-0d4a3cbe6877">

### Expected behavior

My expectation would be, that that pattern matching is valid.

### Actual behavior

Actual behavior is a warning - which leads to build error when having warnings as errors enabled.

### Known workarounds

Using `as` operator for converting doesn't give a warning:

```
var exception = new Exception();
exception.Data.Add("handled", true);

if (exception.Data.Contains("handled") && (exception.Data["handled"] as bool?) == true)
{
    // do something
}
```

<img width="1048" alt="image" src="https://github.com/SonarSource/sonar-dotnet/assets/575109/512bfc9d-bb52-4450-bf98-fadc58c7c843">

### Related information

* C#/VB.NET Plugins version: 9.7.0.75501
* Visual Studio version <-- None ... using Rider 2023.2
* MSBuild / dotnet version: 7.0.400
* SonarScanner for .NET version (if used)
* Operating System: OSX Ventura 13.5

**Comment by zsolt-kolbay-sonarsource** [comment]
Thank you for reporting the issue @kimpenhaus. Confirmed as False Positive.
